pkg/compose/watch.go

```diff
@@ -17,96 +17,89 @@ package compose
 
 func (s *composeService) Watch(ctx context.Context, project *types.Project, services []string, options api.WatchOptions) error {
 	fmt.Fprintln(s.stderr(), "not implemented yet")
 
+	needRebuild := make(chan string)
+	needSync := make(chan api.CopyOptions, 5)
+
 	eg, ctx := errgroup.WithContext(ctx)
-	needRefresh := make(chan string)
 	eg.Go(func() error {
 		clock := clockwork.NewRealClock()
-		debounce(ctx, clock, quietPeriod, needRefresh, func(services []string) {
-			fmt.Fprintf(s.stderr(), "Updating %s after changes were detected\n", strings.Join(services, ", "))
-			imageIds, err := s.build(ctx, project, api.BuildOptions{
-				Services: services,
-			})
-			if err != nil {
-				fmt.Fprintf(s.stderr(), "Build failed")
-			}
-			for i, service := range project.Services {
-				if id, ok := imageIds[service.Name]; ok {
-					service.Image = id
-				}
-				project.Services[i] = service
-			}
-
-			err = s.Up(ctx, project, api.UpOptions{
-				Create: api.CreateOptions{
-					Services: services,
-					Inherit:  true,
-				},
-				Start: api.StartOptions{
-					Services: services,
-					Project:  project,
-				},
-			})
-			if err != nil {
-				fmt.Fprintf(s.stderr(), "Application failed to start after update")
-			}
-		})
+		debounce(ctx, clock, quietPeriod, needRebuild, s.makeRebuildFn(ctx, project))
 		return nil
 	})
 
+	eg.Go(s.makeSyncFn(ctx, project, needSync))
+
 	err := project.WithServices(services, func(service types.ServiceConfig) error {
-		var config DevelopmentConfig
-		if y, ok := service.Extensions["x-develop"]; ok {
-			err := mapstructure.Decode(y, &config)
-			if err != nil {
-				return err
-			}
+		config, err := loadDevelopmentConfig(service, project)
+		if err != nil {
+			return err
 		}
 		if service.Build == nil {
 			return errors.New("can't watch a service without a build section")
 		}
 		context := service.Build.Context
 
-		watcher, err := fsnotify.NewWatcher()
+		ignore, err := watch.LoadDockerIgnore(context)
 		if err != nil {
 			return err
 		}
-		fmt.Println("watching " + context)
-		err = watcher.Add(context)
+
+		watcher, err := watch.NewWatcher([]string{context}, ignore)
+		if err != nil {
+			return err
+		}
+
+		fmt.Fprintf(s.stderr(), "watching %s\n", context)
+		err = watcher.Start()
 		if err != nil {
 			return err
 		}
+
 		eg.Go(func() error {
 			defer watcher.Close() //nolint:errcheck
+		WATCH:
 			for {
 				select {
 				case <-ctx.Done():
 					return nil
-				case event := <-watcher.Events:
-					log.Println("fs event :", event.String())
-					needRefresh <- service.Name
-				case err := <-watcher.Errors:
+				case event := <-watcher.Events():
+					fmt.Fprintf(s.stderr(), "change detected on %s\n", event.Path())
+
+					for src, dest := range config.Sync {
+						if strings.HasPrefix(event.Path(), src) {
+							needSync <- api.CopyOptions{
+								Source:      event.Path(),
+								Destination: fmt.Sprintf("%s:%s/%s", service.Name, dest, event.Path()[len(src):]),
```

Comment:
Ugh, definitely not in-scope for this PR, but we should change `api.CopyOptions` to use full fields for this and do the parsing in the CLI, right now we're doing that all in the backend @ https://github.com/docker/compose/blob/f24d3458c6d05e19519e660e4faa04d47f6db103/pkg/compose/cp.go#L274-L289

Reply:
this indeed is something I had in mind for a while ...